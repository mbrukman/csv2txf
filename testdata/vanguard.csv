"Trade Date","Transaction Type","Investment Name","Symbol","Shares","Principal Amount","Net Amount"
"2011-01-12","Buy","ABC Corp","ABC",300,-123.45,-234.56
"2011-02-01","Sell","ABC Corp","ABC",-300,357.91,245.23
"2011-03-12","Buy","XYZ Inc","XYZ",100,-5432.23,-5678.90
"2011-05-01","Sell","XYZ Inc","XYZ",-100,6892.12,6783.53
