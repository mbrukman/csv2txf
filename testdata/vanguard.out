V042
Acsv2txf
D04/15/2012
^
TD
N321
C1
L1
P300 shares ABC
D01/12/2011
D02/01/2011
$234.56
$245.23
^
TD
N321
C1
L1
P100 shares XYZ
D03/12/2011
D05/01/2011
$5678.90
$6783.53
^