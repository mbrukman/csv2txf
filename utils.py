# Copyright 2012 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Utility methods/classes."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
import sys
from typing import Optional


class Error(Exception):
    pass


class ValueError(Error):

    msg: str

    def __init__(self, msg):
        self.msg = msg

    def __str__(self) -> str:
        return self.msg


class UnimplementedError(Error):

    msg: str

    def __init__(self, msg):
        self.msg = msg

    def __str__(self) -> str:
        return self.msg


def Warning(msg: str):
    sys.stderr.write('warning: %s\n' % msg)


class Transaction:

    desc: Optional[str] = None
    buyDate: Optional[datetime] = None
    buyDateStr: Optional[str] = None
    costBasis: Optional[Decimal] = None
    sellDate: Optional[datetime] = None
    sellDateStr: Optional[str] = None
    saleProceeds: Optional[Decimal] = None
    adjustment: Optional[Decimal] = None
    entryCode: Optional[int] = None

    def __str__(self) -> str:
        data = [
            ('desc:%s', self.desc),
            ('buyDateStr:%s', self.buyDateStr),
            ('costBasis:%.2f', self.costBasis),
            ('sellDateStr:%s', self.sellDateStr),
            ('saleProceeds:%.2f', self.saleProceeds),
            ('adjustment:%.2f', self.adjustment),
            ('entryCode:%d', self.entryCode)
        ]
        formatted_data = [(fmt % value) for (fmt, value) in data if value]
        return ','.join(formatted_data)


def txfDate(date: datetime) -> str:
    """Returns a date string in the TXF format, which is MM/DD/YYYY."""
    return date.strftime('%m/%d/%Y')


def isLongTerm(buy_date: datetime, sell_date: datetime) -> bool:
    # To handle leap years, cannot use a standard number of days, i.e.:
    #   sell_date - buy_date > timedelta(days=365)
    #   - doesn't work for leap years
    #   sell_date - buy_date > timedelta(days=366)
    #   - doesn't work for non-leap years
    if sell_date < buy_date:
        raise ValueError('Sell date before buy date')
    if sell_date.year > buy_date.year + 1:
        return True
    return (sell_date.year == buy_date.year + 1 and
            (sell_date.month > buy_date.month or
             (sell_date.month == buy_date.month and
              sell_date.day > buy_date.day)))
