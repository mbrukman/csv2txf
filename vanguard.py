# Copyright 2012 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Implements Vanguard.

Assumes reconciled transactions, i.e., sell follows buy.

Does not handle:
* dividends
* short sales
* partial lot sales
"""

import csv
from datetime import datetime
from decimal import Decimal
import utils


FIRST_LINE = ','.join(['"Trade Date"', '"Transaction Type"',
                       '"Investment Name"', '"Symbol"', '"Shares"',
                       '"Principal Amount"', '"Net Amount"\n'])


class Vanguard:
    @classmethod
    def name(cls):
        return 'Vanguard'

    @classmethod
    def isBuy(cls, dict):
        return dict['Transaction Type'] == 'Buy'

    @classmethod
    def isSell(cls, dict):
        return dict['Transaction Type'] == 'Sell'

    @classmethod
    def date(cls, dict):
        """Returns date of transaction as datetime object."""
        # Our input date format is YYYY/MM/DD.
        return datetime.strptime(dict['Trade Date'], '%Y-%m-%d')

    @classmethod
    def symbol(cls, dict):
        return dict['Symbol']

    @classmethod
    def investmentName(cls, dict):
        return dict['Investment Name']

    @classmethod
    def numShares(cls, dict):
        shares = int(dict['Shares'])
        if cls.isSell(dict):
            return shares * -1
        else:
            return shares

    @classmethod
    def netAmount(cls, dict):
        amount = Decimal(dict['Net Amount'])
        if cls.isBuy(dict):
            return amount * -1
        else:
            return amount

    @classmethod
    def isFileForBroker(cls, filename):
        with open(filename) as f:
            first_line = f.readline()
            return first_line == FIRST_LINE

    @classmethod
    def parseFileToTxnList(cls, filename, tax_year):
        txns = csv.reader(open(filename), delimiter=',', quotechar='"')
        row_num = 0
        txn_list = []
        names = None
        curr_txn = None
        buy = None
        sell = None
        for row in txns:
            row_num = row_num + 1
            if row_num == 1:
                names = row
                continue

            txn_dict = {}
            for i in range(0, len(names)):
                txn_dict[names[i]] = row[i]

            if cls.isBuy(txn_dict):
                buy = txn_dict
                curr_txn = utils.Transaction()
                curr_txn.desc = '%d shares %s' % (
                    cls.numShares(buy), cls.symbol(buy))
                curr_txn.buyDate = cls.date(txn_dict)
                curr_txn.buyDateStr = utils.txfDate(curr_txn.buyDate)
                curr_txn.costBasis = cls.netAmount(txn_dict)
            elif cls.isSell(txn_dict):
                sell = txn_dict
                # Assume that sells follow the buys, so we can attach this sale to the
                # current buy txn we are processing.
                assert cls.numShares(buy) == cls.numShares(sell)
                assert cls.symbol(buy) == cls.symbol(sell)
                assert cls.investmentName(buy) == cls.investmentName(sell)

                curr_txn.sellDate = cls.date(sell)
                curr_txn.sellDateStr = utils.txfDate(curr_txn.sellDate)
                curr_txn.saleProceeds = cls.netAmount(sell)

                if utils.isLongTerm(curr_txn.buyDate, curr_txn.sellDate):
                    curr_txn.entryCode = 323  # "LT gain/loss - security"
                else:
                    curr_txn.entryCode = 321  # "ST gain/loss - security"

                assert curr_txn.sellDate >= curr_txn.buyDate
                if tax_year and curr_txn.sellDate.year != tax_year:
                    utils.Warning('ignoring txn: "%s" as the sale is not from %d\n' %
                                  (curr_txn.desc, tax_year))
                    continue

                txn_list.append(curr_txn)

                # Clear both the buy and the sell as we have matched them up.
                buy = None
                sell = None
                curr_txn = None

        return txn_list
