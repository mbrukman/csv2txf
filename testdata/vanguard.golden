desc:300 shares ABC,buyDateStr:01/12/2011,costBasis:234.56,sellDateStr:02/01/2011,saleProceeds:245.23,entryCode:321
desc:100 shares XYZ,buyDateStr:03/12/2011,costBasis:5678.90,sellDateStr:05/01/2011,saleProceeds:6783.53,entryCode:321
